### DISCUSSION TITLE
Correct way to pass a sentence

### DISCUSSION THREAD

**Discussion by FionnD** [fastapi/fastapi#4084] [Questions] [answered]
Sorry for the very simple question, i've very new to APIs!

I'm building a very simple application that takes a string (sentence) and returns something. Is it correct just to accept the whole sentence like this? Or should I do something else?

```
@app.get("/{text}")
async def function_endpoint(text: str):
    return {"api": text}
```

**Comment by AlexandreBiguet** [comment]
Hi :)

Are you experiencing a behaviour you didn't expect with this code ? 

If you do, perhaps it is because you are using non ASCII parameters in the `text` URL:

for example, if you want the response to look like 
```json
{ "api": "hello, world" }
```
you should encode the `text` parameter:
```bash
$> curl http://127.0.0.1:8000/hello%2c%20world
```
---

For this part of the question

> Is it correct just to accept the whole sentence like this? Or should I do something else?

Maybe a good starting point could be to have a look at how companies build their APIs, for example [github](https://docs.github.com/en/rest/reference/issues). 

(I may be opinionated here, so other may not agree) 

Most of the time, resources and resource IDs are found in path parameters, whereas options for filtering (sorting, ...) are found as query parameters. 

As an API consumer (the guy that will use your API to build something) I think it makes sense to have required parameters in the path, and optional as query parameters. That said, if you check the [openweathermap](https://openweathermap.org/current) API, you'll see that some query parameters are required. 

To know more about how you can define path and query parameters with fastapi, you can start here: 
* https://fastapi.tiangolo.com/tutorial/path-params/
* https://fastapi.tiangolo.com/tutorial/query-params/

Hope this helps, happy coding :)

**Comment by simondale00** [comment, marked as answer]
Hey @FionnD! I agree with much of @AlexandreBiguet's great advice above. Though if I could make one additional suggestion, it sounds to me like supporting POST instead of GET might be appropriate for your use-case.

With POST, you'll only have a fixed path like `POST /sentence-transformer`, as opposed to an unbound number of paths with `GET /{text}`. Additionally, this pattern may help future-proofing your API. Since `{text}` can be any arbitrary text, it may limit your ability to add other path support in the future.

Additionally, sentences can be _messy_. They may contain non-ascii characters, punctuation, or other characters that may not play nicely in a path. For instance, the following sentences may cause headaches if the text isn't properly URL encoded:

```
this is a sentence with an emoji 😀
this is a sentence/text with a slash which would be interpreted as two parts of a path
this. has. punctuation. which. needs. to. be. URL. encoded
```

However, with POST, you can accept your sentence data in the request body. You won't have the same limitations around encoding or even sentence length.

Good luck!
Simon

**Comment by FionnD** [comment]
Thanks both @simondale00 and @AlexandreBiguet . 

I had been hesitate to use `POST` because from my research, `POST` assumes you are changing something, which isn't the case with ML inference. However I can see the technological advantages of using it. I'll give it a shot, thank you again
